File: src_ios/CnotiMind/rules/RandomNode.m
//
//  RandomNode.m
//  SingleEmotion
//
//  Created by Gonçalo Rodrigues on 9/1/11.
//  Copyright 2011 Universidade de Coimbra. All rights reserved.
//

#import "RandomNode.h"

@implementation RandomNode

- (id) initWithBrainAndParent:(Brain*)aBrain
                       parent:(id)aParent
{
    if (self = [super initWithBrainAndParent:aBrain parent:aParent]) {
        
    }
    
    return self;
}

/**
 Choose a random child node and executes it.
 */
- (void) exec
{
    int count = [[self children] count];
    
    // It must have children nodes to run
    if(count == 0)
    {
        return;
    }
    
    int n = arc4random() % count;
    
    RuleNode* node = [[self children] objectAtIndex:n-1];
    [node exec];     
}


- (void) exec:(NSMutableDictionary*)aVariables
{
    int count = [[self children] count];
    
    // It must have children nodes to run
    if(count == 0)
    {
        return;
    }
    
    int n = arc4random() % count;
    
    RuleNode* node = [[self children] objectAtIndex:n-1];
    [node exec:aVariables]; 

}

- (NSString*) info:(int)aDepth
{
    NSString* info = [NSString stringWithFormat:@""];
    NSString* space = [self space:aDepth];
    
    info = [info stringByAppendingFormat:@"%@ Random", space];
    
    return info;  
}


- (void) dealloc
{
    [super dealloc];
}


@end

File: src_qt/rules/RandomNode.cpp
#include "RandomNode.h"
#include "../Brain.h"

namespace CnotiMind
{

	RandomNode::RandomNode(Brain* brain, QObject *parent) :
		RuleNode(brain, parent)
	{
	}

	/**
		Choose a random child node and executes it.
	*/
	void RandomNode::exec()
	{
		int count = children().size();

		// It must have children nodes to run
		if(count == 0)
		{
			return;
		}

		int n = qrand() % count;

		RuleNode* node = (RuleNode* ) children().value( n );

		node->exec();

	}

	void RandomNode::exec( QHash<QString, QString>& variables )
	{
		int count = children().size();

		// It must have children nodes to run
		if(count == 0)
		{
			return;
		}

		int n = qrand() % count;

		RuleNode* node = ( RuleNode* ) children().at( n );
		node->exec( variables );
	}

	QString RandomNode::info( int depth ) const
	{
		return space( depth ) + "Random";
	}

	RandomNode *RandomNode::fromXML( const QString &qName, const QXmlAttributes &atts, Brain* brain, QObject* parent )
	{
        Q_UNUSED(atts);

		if( qName.compare( "Random", Qt::CaseInsensitive ) == 0 )
		{
			return new RandomNode(brain, parent);
		}
		return NULL;
	}
}
